File: metals/src/main/scala/scala/meta/internal/metals/DefinitionResult.scala
package scala.meta.internal.metals

import java.util
import java.util.Collections

import scala.meta.internal.metals.MetalsEnrichments._
import scala.meta.internal.semanticdb.Scala.Symbols
import scala.meta.internal.semanticdb.TextDocument
import scala.meta.io.AbsolutePath

import org.eclipse.lsp4j.Location

case class DefinitionResult(
    locations: util.List[Location],
    symbol: String,
    definition: Option[AbsolutePath],
    semanticdb: Option[TextDocument],
    querySymbol: String,
) {
  def isEmpty: Boolean = locations.isEmpty()
  def ++(other: DefinitionResult): DefinitionResult = DefinitionResult(
    (locations.asScala ++ other.locations.asScala).asJava,
    symbol,
    definition,
    semanticdb,
    querySymbol,
  )
}

object DefinitionResult {
  def empty(symbol: String): DefinitionResult =
    DefinitionResult(Collections.emptyList(), symbol, None, None, symbol)
  def empty: DefinitionResult = empty(Symbols.None)
}

File: mtags-interfaces/src/main/java/scala/meta/pc/DefinitionResult.java
package scala.meta.pc;

import java.util.Collections;
import java.util.List;
import org.eclipse.lsp4j.Location;

public interface DefinitionResult {
  static DefinitionResult empty() {
    return new DefinitionResult() {
      @Override
      public String symbol() {
        return "";
      }

      @Override
      public List<Location> locations() {
        return Collections.emptyList();
      }
    };
  }

  String symbol();

  List<Location> locations();

  /**
   * A resolved definition has full location information, while an unresolved definition may only
   * have a symbol and a uri, but no position. Currently, the only case where we return unresolved
   * definitions is when we are decompiling code from a jar file. The PC does not have access to a
   * decompiler, so it relies on the client to decompile and locate the symbol definition inside the
   * decompiled code.
   */
  default boolean isResolved() {
    return true;
  }

  /**
   * The simple names of the resolved method's parameters, in declaration order, or empty if the
   * symbol isn't a method or this information isn't available.
   */
  default List<String> parameterNames() {
    return Collections.emptyList();
  }

  /**
   * The type names of the resolved method's parameters, in declaration order, or empty if the
   * symbol isn't a method or this information isn't available.
   */
  default List<String> parameterTypeNames() {
    return Collections.emptyList();
  }
}
